src/vcpkg/portfileprovider.cpp

```diff
@@ -10,10 +11,309 @@
 #include <vcpkg/registries.h>
 #include <vcpkg/sourceparagraph.h>
 
+#include <functional>
+#include <map>
+
 using namespace vcpkg;
 
 namespace vcpkg
 {
+    OverlayPortIndexEntry::OverlayPortIndexEntry(OverlayPortKind kind, const Path& directory)
+        : m_kind(kind), m_directory(directory), m_loaded_ports()
+    {
+        if (m_kind == OverlayPortKind::Port)
+        {
+            Checks::unreachable(VCPKG_LINE_INFO);
+        }
+    }
+
+    OverlayPortIndexEntry::OverlayPortIndexEntry(OverlayPortIndexEntry&&) = default;
+
+    OverlayPortKind OverlayPortIndexEntry::determine_kind(const ReadOnlyFilesystem& fs)
+    {
+        if (m_kind == OverlayPortKind::Unknown)
+        {
+            if (!m_loaded_ports.empty())
+            {
+                Checks::unreachable(VCPKG_LINE_INFO, "OverlayPortKind::Unknown empty cache constraint violated");
+            }
+
+            auto maybe_scfl = Paragraphs::try_load_port(fs, PortLocation{m_directory}).maybe_scfl;
+            if (auto scfl = maybe_scfl.get())
+            {
+                if (scfl->source_control_file)
+                {
+                    // succeeded in loading it, so this must be a port
+                    m_kind = OverlayPortKind::Port;
+                    auto name_copy = scfl->to_name(); // copy name before moving maybe_scfl
+                    m_loaded_ports.emplace(name_copy, std::move(maybe_scfl));
+                }
+                else
+                {
+                    // the directory didn't look like a port at all, consider it an overlay-port-dir
+                    m_kind = OverlayPortKind::Directory;
+                }
```

Comment:
I would expect that if `Paragraphs::try_load_port` is successful it loaded a port... 

Also I find it not so nice that `m_kind` is initialized in the constructor (but not for all cases) and here again. Imho it would be nicer to determine that beforehand and maybe even have distinct types.

Reply:
> I would expect that if `Paragraphs::try_load_port` is successful it loaded a port...

You want `try_load_port_required` for that. "There was no error but there is also no port of that name" is a valid success condition: https://github.com/microsoft/vcpkg-tool/blob/d176cbeeb07c9d8fe8ae3953fab985d53e8ec483/include/vcpkg/paragraphs.h#L38-L41

> Also I find it not so nice that m_kind is initialized in the constructor (but not for all cases) and here again. Imho it would be nicer to determine that beforehand and maybe even have distinct types.

The problem is that determining this answer requires hitting the disk and we have no reason to do that if we don't happen to consider this `overlay-ports` entry in the first place.